File: Homework/Manh/lesson-1/SortArrayByParityII.java
class Solution {
    public int[] sortArrayByParityII(int[] nums) {
        int odd = 1;
        int even = 0;
        while(odd<nums.length && even<nums.length){
            if(nums[even] % 2 != 0){
                swap(odd, even, nums);
                odd = odd + 2;
            } else {
                even = even + 2;
            }
        }
        return nums;
    }
    
    private void swap(int x, int y, int[] nums) {
        if(x==y) return;
        nums[x] = nums[x] + nums[y];
        nums[y] = nums[x] - nums[y];
        nums[x] = nums[x] - nums[y];
    }
}

File: Homework/Khang_Nguyen/sort-array-by-parity-ii.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
function swap(start, end, arr) {
    const temp = arr[start]
    arr[start] = arr[end]
    arr[end] = temp
}
var sortArrayByParityII = function(nums) {
  const len = nums.length
  let nextOddIndex = 1
  let nextEvenIndex = 0
  
  while (nextEvenIndex < len || nextOddIndex < len - 1) {
      const isEvenPass = nums[nextEvenIndex] % 2 === 0
      const isOddPass = nums[nextOddIndex] % 2 === 1

      if (isEvenPass) {
          nextEvenIndex += 2
      } else if (isOddPass) {
          nextOddIndex += 2
      } else if (!isEvenPass && !isOddPass) {
          swap(nextEvenIndex, nextOddIndex, nums)
          nextEvenIndex += 2
          nextOddIndex += 2
      }
  }
  return nums
};